File: redhawk-codegen/redhawk/codegen/jinja/cpp/properties/templates/properties.cpp
/*#
 * This file is protected by Copyright. Please refer to the COPYRIGHT file 
 * distributed with this source distribution.
 * 
 * This file is part of REDHAWK core.
 * 
 * REDHAWK core is free software: you can redistribute it and/or modify it 
 * under the terms of the GNU Lesser General Public License as published by the 
 * Free Software Foundation, either version 3 of the License, or (at your 
 * option) any later version.
 * 
 * REDHAWK core is distributed in the hope that it will be useful, but WITHOUT 
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License 
 * for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License 
 * along with this program.  If not, see http://www.gnu.org/licenses/.
 #*/
/*{% macro initsequence(prop) %}*/
//%   if prop is structsequence
${prop.cppname}.resize(${prop.cppvalues|length});
//%     for value in prop.cppvalues
//%       set index = loop.index0
//%       for field in prop.structdef.fields
${prop.cppname}[${index}].${field.cppname} = ${value[field.identifier]};
//%       endfor
//%     endfor
//%   else
// Set the sequence with its initial values
//%     for value in prop.cppvalues
${prop.cppname}.push_back(${value});
//%     endfor
//%   endif
/*{%- endmacro %}*/

/*{% macro addproperty(prop) %}*/
addProperty(${prop.cppname},
//% if prop.cppvalues
            ${prop.cppname},
//% elif prop.cppvalue
            ${prop.cppvalue},
//% endif
            "${prop.identifier}",
            "${prop.name}",
            "${prop.mode}",
            "${prop.units}",
            "${prop.action}",
            "${prop.kinds|join(',')}");
/*{%- endmacro %}*/

/*{% macro initializestructseq(prop) %}*/
/*{% for value in prop.cppvalues %}*/
    {
        ${prop.structdef.cpptype} __tmp;
/*{%   for key in value %}*/
/*{%     for field in prop.structdef.fields %}*/
/*{%       if field.identifier == key %}*/
/*{%         if field is simple %}*/
        __tmp.${field.cppname} = ${value[key]};
/*{%         elif field is simplesequence%}*/
/*{%           for fieldvalue in value[key] %}*/
        __tmp.${field.cppname}.push_back(${fieldvalue});
/*{%           endfor %}*/
/*{%         endif %}*/
/*{%       endif %}*/
/*{%     endfor %}*/
/*{%   endfor %}*/
        ${prop.cppname}.push_back(__tmp);
    }
/*{% endfor %}*/
/*{%- endmacro %}*/

/*{% macro structdef(struct) %}*/
/*{%   for field in struct.fields if field.enums %}*/
/*{%     if loop.first %}*/
namespace enums {
    // Enumerated values for ${struct.identifier}
    namespace ${struct.cppname} {
/*{%     endif %}*/
        ${enumvalues(field)|indent(8)}
/*{%     if loop.last %}*/
    }
}

/*{%     endif %}*/
/*{%   endfor %}*/
struct ${struct.cpptype}${' : public '+struct.baseclass if struct.baseclass} {
    ${struct.cpptype} ()${' : '+struct.baseclass+'()' if struct.baseclass}
    {
/*{% for field in struct.fields %}*/
/*{%   if not field.inherited %}*/
/*{%     if field is simple and field.cppvalue %}*/
        ${field.cppname} = ${field.cppvalue};
/*{%     elif field is simplesequence and field.cppvalues %}*/
/*{%       for value in field.cppvalues %}*/
        ${field.cppname}.push_back(${value});
/*{%       endfor %}*/
/*{%     endif %}*/
/*{%   endif %}*/
/*{% endfor %}*/
    }

    static std::string getId() {
        return std::string("${struct.identifier}");
    }

    static const char* getFormat() {
        return "${struct.format}";
    }
/*{% for field in struct.fields if not field.inherited %}*/
/*{%   if loop.first %}*/

/*{%   endif %}*/
/*{%   if field.isOptional %}*/
/*{%     if field is simplesequence %}*/
    optional_property<${field.cpptype} > ${field.cppname};
/*{%     else %}*/
    optional_property<${field.cpptype}> ${field.cppname};
/*{%     endif %}*/
/*{%   else %}*/
    ${field.cpptype} ${field.cppname};
/*{%   endif %}*/
/*{% endfor %}*/
};

inline bool operator>>= (const CORBA::Any& a, ${struct.cpptype}& s) {
    CF::Properties* temp;
    if (!(a >>= temp)) return false;
    const redhawk::PropertyMap& props = redhawk::PropertyMap::cast(*temp);
/*{% for field in struct.fields %}*/
    if (props.contains("${field.identifier}")) {
/*{% if field.isOptional %}*/
        if (!(ossie::any::isNull(props["${field.identifier}"]))) {
/*{%   if field is simple %}*/
/*{%     if field.type == 'char' %}*/
            CORBA::Char tmp;
/*{%     else %}*/
            ${field.cpptype} tmp;
/*{%     endif %}*/
/*{%     set extractName = 'tmp' %}*/
            if (!(props["${field.identifier}"] >>= ${cpp.extract(extractName, field.type, field.iscomplex)})) return false;
/*{%   elif field is simplesequence %}*/
            ${field.cpptype} tmp;
            if (!(props["${field.identifier}"] >>= tmp)) return false;
/*{%   endif %}*/
            s.${field.cppname} = tmp;
        } else {
            s.${field.cppname}.reset();
        }
/*{% else %}*/
/*{%   if field is simple %}*/
/*{%     if field.type == 'char' %}*/
/*{%       set extractName = 'temp_char' %}*/
        CORBA::Char temp_char;
/*{%     else %}*/
/*{%       set extractName = 's.'+field.cppname %}*/
/*{%     endif %}*/
        if (!(props["${field.identifier}"] >>= ${cpp.extract(extractName, field.type, field.iscomplex)})) return false;
/*{%     if field.type == 'char' %}*/
        s.${field.cppname} = temp_char;
/*{%     endif %}*/
/*{%   elif field is simplesequence %}*/
        if (!(props["${field.identifier}"] >>= s.${field.cppname})) return false;
/*{%   endif %}*/
/*{% endif %}*/
    }
/*{% endfor %}*/
    return true;
}

inline void operator<<= (CORBA::Any& a, const ${struct.cpptype}& s) {
    redhawk::PropertyMap props;
/*{% for field in struct.fields %}*/
/*{%   if field.isOptional %}*/
    if (s.${field.cppname}.isSet()) {
/*{%     if field is simple %}*/
        props["${field.identifier}"] = ${cpp.insert('*(s.'+field.cppname+')', field.type, field.iscomplex)};
/*{%     elif field is simplesequence %}*/
        props["${field.identifier}"] = *(s.${field.cppname});
/*{%     endif %}*/
    }
/*{%   else %}*/ 
/*{%     if field is simple %}*/
    props["${field.identifier}"] = ${cpp.insert('s.'+field.cppname, field.type, field.iscomplex)};
/*{%     elif field is simplesequence %}*/
    props["${field.identifier}"] = s.${field.cppname};
/*{%     endif %}*/
/*{%   endif %}*/
/*{% endfor %}*/
    a <<= props;
}

inline bool operator== (const ${struct.cpptype}& s1, const ${struct.cpptype}& s2) {
/*{% for field in struct.fields %}*/
    if (s1.${field.cppname}!=s2.${field.cppname})
        return false;
/*{% endfor %}*/
    return true;
}

inline bool operator!= (const ${struct.cpptype}& s1, const ${struct.cpptype}& s2) {
    return !(s1==s2);
}
/*{%- endmacro %}*/

/*{% macro enumvalues(prop) %}*/
// Enumerated values for ${prop.identifier}
namespace ${prop.cppname} {
/*{% for enum in prop.enums %}*/
    static const ${prop.cpptype} ${enum.cpplabel} = ${enum.cppvalue};
/*{% endfor %}*/
}
/*{%- endmacro %}*/

File: redhawk-codegen/redhawk/codegen/model/properties.py
#
# This file is protected by Copyright. Please refer to the COPYRIGHT file
# distributed with this source distribution.
#
# This file is part of REDHAWK core.
#
# REDHAWK core is free software: you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option) any
# later version.
#
# REDHAWK core is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see http://www.gnu.org/licenses/.
#

import ossie.parsers.prf

from redhawk.codegen.utils import strenum

Kinds = strenum('execparam', 'configure', 'allocation', 'factoryparam', 'test', 'event', 'message')
Modes = strenum('writeonly', 'readonly', 'readwrite')
Actions = strenum('external', 'eq', 'ge', 'gt', 'le', 'lt', 'ne')

class Property(object):

    def __init__(self, xml):
        self.xml = xml

    def kinds(self):
        kinds = self.kinds_()
        if len(kinds) > 0:
            return set(str(k.kindtype) for k in kinds)
        else:
            return set((Kinds.CONFIGURE,))

    def identifier(self):
        return self.xml.id_

    def hasName(self):
        return self.xml.name is not None

    def name(self):
        if self.xml.name:
            return self.xml.name
        else:
            return self.xml.id_

    def mode(self):
        if not self.xml.mode:
            return Modes.READWRITE
        else:
            return str(self.xml.mode)

    def hasDescription(self):
        return self.xml.description is not None

    def description(self):
        return self.xml.description

    def isConfigure(self):
        return Kinds.CONFIGURE in self.kinds()

    def isAllocation(self):
        return Kinds.ALLOCATION in self.kinds()

    def isFactoryParam(self):
        return Kinds.FACTORYPARAM in self.kinds()

    def isTest(self):
        return Kinds.TEST in self.kinds()

    def isEvent(self):
        return Kinds.EVENT in self.kinds()

    def isMessage(self):
        return Kinds.MESSAGE in self.kinds()

class _Simple(object):
    def kinds_(self):
        return self.xml.kind

    def isStruct(self):
        return False

    def type(self):
       return self.xml.type_

    def isComplex(self):
        if self.xml.complex.lower() == "true":
            return True
        else:
            return False

    def isOptional(self):
        if self.xml.optional.lower() == "true":
            return True
        else:
            return False

    def action(self):
        if not self.xml.action or not self.xml.action.type_:
            return Actions.EXTERNAL
        else:
            return str(self.xml.action.type_)

    def range(self):
        if self.xml.range_:
            return self.xml.range_.min, self.xml.range_.max
        else:
            return None

    def hasUnits(self):
        return self.xml.units is not None

    def units(self):
        if self.xml.units:
            return self.xml.units
        else:
            return ""

class _Struct(object):
    def isStruct(self):
        return True

    def kinds_(self):
        return self.xml.configurationkind    

    def hasUnits(self):
        return False

    def units(self):
        return ""

    def action(self):
        return Actions.EXTERNAL

class _Single(object):
    def isSequence(self):
        return False

class _Sequence(object):
    def isSequence(self):
        return True

class SimpleProperty(Property, _Simple, _Single):
    def hasValue(self):
        return self.xml.value is not None

    def value(self):
        return self.xml.value

    def hasEnumerations(self):
        return bool(self.xml.enumerations)

    def enumerations(self):
        if self.xml.enumerations:
            return [(e.label, e.value) for e in self.xml.enumerations.enumeration]
        else:
            return None

class SimpleSequenceProperty(Property, _Simple, _Sequence):
    def hasValue(self):
        return self.xml.values is not None

    def value(self):
        if self.hasValue():
            return self.xml.values.value
        return None

class StructProperty(Property, _Struct, _Single):
    def type(self):
        return 'struct'

    def fields(self):
        f = [SimpleProperty(s) for s in self.xml.simple]
        f += [SimpleSequenceProperty(s) for s in self.xml.simplesequence]
        return f

    def hasValue(self):
        for field in self.fields():
            if field.value is not None:
                return True
        return False

    def value(self):
        return dict((s.name(), s.value()) for s in self.fields())

class StructSequenceProperty(Property, _Struct, _Sequence):
    def type(self):
        return 'structsequence'

    def struct(self):
        return StructProperty(self.xml.struct)

    def hasValue(self):
        return len(self.xml.structvalue) > 0

    def value(self):
        structref = self.struct()
        base = structref.value()
        idmap = dict((s.identifier(), s.name()) for s in structref.fields())
        return [self.mapvalue_(base, idmap, v) for v in self.xml.structvalue]

    def mapvalue_(self, base, mapping, structval):
        value = base.copy()
        value.update((v.refid, v.value) for v in structval.simpleref)
        value.update((v.refid, v.values.value) for v in structval.simplesequenceref)
        return value

def parse(prfFile):
    prf = ossie.parsers.prf.parse(prfFile)
    props = [SimpleProperty(p) for p in prf.simple]
    props += [SimpleSequenceProperty(p) for p in prf.simplesequence]
    props += [StructProperty(p) for p in prf.struct]
    props += [StructSequenceProperty(p) for p in prf.structsequence]
    return props
